crypto/keys/pem_signer_factory_test.go

```diff
@@ -107,3 +113,101 @@ func TestPEMSignerFactoryNewSigner(t *testing.T) {
 		}
 	}
 }
+
+func TestPEMSignerFactoryGenerate(t *testing.T) {
+	for _, test := range []struct {
+		name    string
+		keygen  *keyspb.Generator
+		tree    *trillian.Tree
+		wantErr bool
+	}{
+		{
+			name: "RSA",
+			keygen: &keyspb.Generator{
+				Params: &keyspb.Generator_RsaParams{
+					RsaParams: &keyspb.Generator_RSA{},
+				},
+			},
+			tree: &trillian.Tree{
+				SignatureAlgorithm: sigpb.DigitallySigned_RSA,
+			},
+		},
+		{
+			name: "ECDSA",
+			keygen: &keyspb.Generator{
+				Params: &keyspb.Generator_EcdsaParams{
+					EcdsaParams: &keyspb.Generator_ECDSA{},
+				},
+			},
+			tree: &trillian.Tree{
+				SignatureAlgorithm: sigpb.DigitallySigned_ECDSA,
+			},
+		},
+		{
+			name: "Replace existing key",
+			keygen: &keyspb.Generator{
+				Params: &keyspb.Generator_EcdsaParams{
+					EcdsaParams: &keyspb.Generator_ECDSA{},
+				},
+			},
+			tree: &trillian.Tree{
+				PrivateKey: marshalAny(&keyspb.PEMKeyFile{
+					Path: "non-existent.pem",
+				}),
+				PublicKey: &keyspb.PublicKey{
+					Der: []byte("foo"),
+				},
+				SignatureAlgorithm: sigpb.DigitallySigned_ECDSA,
+			},
+		},
+		{
+			name: "Mismatched Generator.Params and tree.SignatureAlgorithm",
+			keygen: &keyspb.Generator{
+				Params: &keyspb.Generator_EcdsaParams{
+					EcdsaParams: &keyspb.Generator_ECDSA{},
+				},
+			},
+			tree: &trillian.Tree{
+				SignatureAlgorithm: sigpb.DigitallySigned_RSA,
+			},
+			wantErr: true,
+		},
+	} {
+		ctx := context.Background()
+		var sf PEMSignerFactory
+
+		newTree, err := sf.Generate(ctx, test.tree, test.keygen)
```

Comment:
As it stands, `Generate()` could be modifying unrelated tree fields and this test would be none the wiser. I think the better approach here is to change Generate's return, as stated in a following comment.